dhis-2/dhis-e2e-test/src/test/java/org/hisp/dhis/tracker/acl/TrackedEntityInstanceAclReadTests.java

```diff
@@ -0,0 +1,539 @@
+    }
+
+    /**
+     * Asserts that the enrollment follows the expectations.
+     *
+     * @param user       the user(username) we are testing as
+     * @param enrollment the enrollment we are testing
+     * @param tei        the tei wrapped around the enrollment
+     */
+    private void assertEnrollment( User user, JsonObject enrollment, JsonObject tei )
+    {
+        String program = enrollment.get( _PROGRAM ).getAsString();
+        String orgUnit = enrollment.get( _OU ).getAsString();
+
+        if ( !user.hasAllAuthority() )
+        {
+            assertStringIsInWhitelistOrNotInBlacklist( user.getDataRead().get( "programs" ),
+                user.getNoDataRead().get( "programs" ), program );
+        }
+        assertSameValueForProperty( tei, enrollment, _TEI );
+        assertWithinOuScope( user.getScopes(), Lists.newArrayList( orgUnit ) );
+        assertNotDeleted( enrollment );
+
+        assertTrue( enrollment.has( _EVENTS ) );
+
+        enrollment.get( _EVENTS ).getAsJsonArray()
+            .forEach( eventJson -> assertEvent( user, eventJson.getAsJsonObject(), enrollment ) );
+    }
+
+    /**
+     * Asserts that the event follows the expectations.
+     *
+     * @param user       the user(username) we are testing as
+     * @param event      the event we are testing
+     * @param enrollment the enrollment wrapped around the event
+     */
+    private void assertEvent( User user, JsonObject event, JsonObject enrollment )
+    {
+        String programStage = event.get( _PROGRAMSTAGE ).getAsString();
+        String orgUnit = event.get( _OU ).getAsString();
+
+        if ( !user.hasAllAuthority() )
+        {
+            assertStringIsInWhitelistOrNotInBlacklist( user.getDataRead().get( "programStages" ),
+                user.getNoDataRead().get( "programStages" ), programStage );
+        }
+        assertWithinOuScope( user.getScopes(), Lists.newArrayList( orgUnit ) );
+        assertSameValueForProperty( enrollment, event, _ENROLLMENT );
+        assertSameValueForProperty( enrollment, event, _TEI );
+        assertNotDeleted( event );
+    }
+
+    private void assertAttribute( User user, JsonObject attribute )
+    {
+        String attributeUid = attribute.get( _ATTRIBUTE ).getAsString();
+
+        // NoDataRead includes all attributes with metadata read, so we use NoDataRead to check access,
+        // instead of DataRead, since there is no DataRead for attributes.
+
+        if ( !user.hasAllAuthority() )
+        {
+            assertStringIsInWhitelistOrNotInBlacklist( user.getNoDataRead().get( "trackedEntityAttributes" ),
+                Lists.newArrayList(), attributeUid );
+        }
+    }
+
+    /**
+     * Asserts that the given JsonObject does not have a property "deleted" that is true.
+     *
+     * @param object the object to check
+     */
+    private void assertNotDeleted( JsonObject object )
+    {
+        assertTrue( object.has( _DELETED ) && !object.get( _DELETED ).getAsBoolean() );
```

Comment:
In all the asserts, provide a meaningful message as well for better reporting. For example: 
`assertTrue(String.format('Object %s with id %s was marked as deleted', .. .. ), object.has( _DELETED ) && !object.get( _DELETED ).getAsBoolean())`